### PR TITLE
[Merged by Bors] - [Fixes #6059] ``Entity``'s “ID” should be named “index” instead

### LINKED ISSUE #6059: `Entity`'s “ID” should be named “index” instead
In Bevy ECS, an `Entity` is just an identifier (ID for short) that is implemented as a generational index: a data structure containing both an index and a generation.

There is a problem however: the entity's index is actually called “ID”, both in documentation and in the APIs.

## How can Bevy's documentation be improved?

The solution would be to change all incorrect occurrences of “ID” to “index”.

[`Entity`] seems to be the only item that needs to be fixed. In particular:
- struct level documentation
- `id` struct field
- `id` method and its documentation

`EntityCommands::id` and `EntityMut::id` are correct, they must not be changed.

**Remember to add the `C-Breaking-Change` label to your PR.**

[`Entity`]: https://dev-docs.bevyengine.org/bevy/ecs/entity/struct.Entity.html

### PR DESCRIPTION
**Pull request by Edwox**
# Objective

Fixes #6059, changing all incorrect occurrences of ``id`` in the ``entity`` module to ``index``:

* struct level documentation,
* ``id`` struct field,
* ``id`` method and its documentation.

## Solution

Renaming and verifying using CI. 

## Migration Guide

The `Entity::id()` method was renamed to `Entity::index()`.

### CODE REVIEW

**Comment by Edwox** [comment]
Could someone help me apply the label specified in #6059?

File: crates/bevy_ecs/src/entity/mod.rs [hunk 1]
```diff
@@ -164,8 +164,11 @@ impl Entity {
     ///     }
     /// }
     /// ```
-    pub const fn from_raw(id: u32) -> Entity {
-        Entity { id, generation: 0 }
+    pub const fn from_raw(index: u32) -> Entity {
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/entity/mod.rs:167]: The doc of this method should be update too: there is multiple mention to "id".

File: crates/bevy_ecs/src/entity/mod.rs [hunk 2]
```diff
@@ -107,7 +107,7 @@ type IdCursor = isize;
 #[derive(Clone, Copy, Hash, Eq, Ord, PartialEq, PartialOrd)]
 pub struct Entity {
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/entity/mod.rs:108]: The doc must be updated: there is mention to "id".
```
/// The identifier is implemented using a [generational index]: a combination of an ID and a generation.
```

File: crates/bevy_ecs/src/entity/mod.rs [hunk 3]
```diff
@@ -195,7 +198,7 @@ impl Entity {
     /// specific snapshot of the world, such as when serializing.
     #[inline]
     pub const fn id(self) -> u32 {
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/entity/mod.rs:200]: The doc mention "id".

Comment (tguichaoua) [inline crates/bevy_ecs/src/entity/mod.rs:200]:
```suggestion
    pub const fn index(self) -> u32 {
```

File: crates/bevy_ecs/src/entity/mod.rs [hunk 4]
```diff
@@ -242,11 +245,16 @@ impl<'a> Iterator for ReserveEntitiesIterator<'a> {
     fn next(&mut self) -> Option<Self::Item> {
         self.id_iter
```
Comment (tguichaoua) [inline crates/bevy_ecs/src/entity/mod.rs:246]:
```rust
    // Reserved IDs formerly in the freelist to hand out.
    id_iter: std::slice::Iter<'a, u32>,
    // New Entity IDs to hand out, outside the range of meta.len().
    id_range: std::ops::Range<u32>,
```

`id_iter` and `id_range` can be renamed `index_iter` and `index_range`

**Review by tguichaoua — CHANGES_REQUESTED** [review]
There is few places where "id" should be change into "index".

**Comment by Nilirad** [comment]
> Could someone help me apply the label specified in #6059? It was the `C-Breaking-Change` label.

Done. (my bad, I forgot that only organization members can tag issues and PRs)

File: crates/bevy_ecs/src/entity/mod.rs (hunk 3 above)
Comment (cart) [reply crates/bevy_ecs/src/entity/mod.rs:200]: This is definitely a more invasive change, as the rename touches a lot of code. But I think its still worth doing, as it provides the right mental model for the field value. I recommend doing the rename with rust-analyzer, as merging this suggestion as-is will break a lot of code.

**Comment by Edwox** [comment]
> There is few places where "id" should be change into "index".

My plan was to do the changes incrementally.  So I started off by just changing the ``id`` in the ``entity`` struct and all resulting dependency issues. The plan was then to start working on the documentation and so on.

> > Could someone help me apply the label specified in #6059? It was the `C-Breaking-Change` label.
> 
> Done. (my bad, I forgot that only organization members can tag issues and PRs)

Thank you! :)

**Comment by Nilirad** [comment]
Looks like there's some code that is only compiled in wasm builds. It wasn't changed probably because rust-analyzer checks for the default target.

If you do `cargo check -p bevy_ecs --all-targets` you may be able to reproduce the errors locally (`-p bevy_ecs` only restricts the check to that crate).

**Comment by Edwox** [comment]
> Looks like there's some code that is only compiled in wasm builds. It wasn't changed probably because rust-analyzer checks for the default target.
> 
> If you do `cargo check -p bevy_ecs --all-targets` you may be able to reproduce the errors locally (`-p bevy_ecs` only restricts the check to that crate).

Thank you! I believe that the issue should be resolved now. :smiley:

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 5]
```diff
@@ -225,7 +225,7 @@ impl ComponentSparseSet {
                 #[cfg(not(debug_assertions))]
                 let idx = swapped_entity;
                 #[cfg(debug_assertions)]
-                let idx = swapped_entity.id();
+                let idx = swapped_entity.index();
                 *self.sparse.get_mut(idx).unwrap() = dense_index as u32;
```
Comment (xgbwei) [inline crates/bevy_ecs/src/storage/sparse_set.rs:229]: `idx` appears again here.

**Review by xgbwei — CHANGES_REQUESTED** [review]
A few variables named `idx` in sparse_set.rs need to be updated.

**Review by Edwox — COMMENTED** [review]
okay, I agree with these changes. I will get right on it. 😄

**Comment by Edwox** [comment]
> I'm in favor of this change, and this is well-done. Can you please resolve the comments as you address them to make it easier for reviewers to track what needs to be done?

Great, thank you! I think all comments should be resolved now.

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 6]
```diff
@@ -65,8 +65,8 @@ impl DynamicScene {
         let type_registry = registry.read();
 
         for scene_entity in &self.entities {
-            // Fetch the entity with the given entity id from the `entity_map`
-            // or spawn a new entity with a transiently unique id if there is
+            // Fetch the entity with the given entity index from the `entity_map`
+            // or spawn a new entity with a transiently unique index if there is
```
Comment (devil-ira) [inline crates/bevy_scene/src/dynamic_scene.rs:69]: `id` is correct here.

File: crates/bevy_scene/src/dynamic_scene_builder.rs [hunk 7]
```diff
@@ -92,20 +92,20 @@ impl<'w> DynamicSceneBuilder<'w> {
...
                 components: Vec::new(),
             };
 
-            for component_id in self.world.entity(entity).archetype().components() {
+            for component_index in self.world.entity(entity).archetype().components() {
                 let reflect_component = self
                     .world
                     .components()
-                    .get_info(component_id)
+                    .get_info(component_index)
```
Comment (devil-ira) [inline crates/bevy_scene/src/dynamic_scene_builder.rs:108]: `component_id` is correct here.

**Review by devil-ira — CHANGES_REQUESTED** [review]

**Review by Edwox — COMMENTED** [review]
Of course, my bad! Will fix this right away.

**Review by alice-i-cecile — APPROVED** [review]
Rebase LGTM.

**Comment by devil-ira** [comment]
Looks like there are still some calls to `.id()` left that need to be changed to `.index()`

**Comment by Edwox** [comment]
> Looks like there are some new calls to `.id()` that need to be changed to `.index()` that came in from the merge.

Attempting to fix it right now.

**Comment by alice-i-cecile** [comment]
Can you please add the following to your PR description?

## Migration Guide

The `Entity::id()` method was renamed to `Entity::index()`.

bors r+

**Comment by Edwox** [comment]
The description should now be updated.

**Pull request closed without merging**